# 1. Dataset Loading and Initial Assessment

The original employee dataset was loaded into a pandas DataFrame for data-cleaning and quality assessment.

The initial assessment focused on understanding the dataset structure, dimensions, column names, data types, missing values, and duplicate records before applying cleaning operations.

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
file_path = "../dataset/Messy_Employee_dataset.csv"

df = pd.read_csv(file_path)

print("Dataset loaded successfully.")
print("Shape:", df.shape)

Dataset loaded successfully.
Shape: (1020, 12)


In [3]:
df.head()

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work
0,EMP1000,Bob,Davis,25.0,DevOps-California,Active,4/2/2021,59767.65,bob.davis@example.com,-1651623197,Average,True
1,EMP1001,Bob,Brown,NaN,Finance-Texas,Active,7/10/2020,65304.66,bob.brown@example.com,-1898471390,Excellent,True
2,EMP1002,Alice,Jones,NaN,Admin-Nevada,Pending,12/7/2023,88145.90,alice.jones@example.com,-5596363211,Good,True
3,EMP1003,Eva,Davis,25.0,Admin-Nevada,Inactive,11/27/2021,69450.99,eva.davis@example.com,-3476490784,Good,True
4,EMP1004,Frank,Williams,25.0,Cloud Tech-Florida,Active,1/5/2022,109324.61,frank.williams@example.com,-1586734256,Poor,False


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   str    
 1   First_Name         1020 non-null   str    
 2   Last_Name          1020 non-null   str    
 3   Age                809 non-null    float64
 4   Department_Region  1020 non-null   str    
 5   Status             1020 non-null   str    
 6   Join_Date          1020 non-null   str    
 7   Salary             996 non-null    float64
 8   Email              1020 non-null   str    
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   str    
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), str(8)
memory usage: 88.8 KB


In [5]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Employee_ID,1020,1020,EMP1000,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
First_Name,1020,8,Frank,142,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Last_Name,1020,8,Brown,148,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Age,809.0,NaN,NaN,NaN,32.484549,5.65686,25.0,25.0,30.0,40.0,40.0
Department_Region,1020,36,HR-Florida,41,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Status,1020,3,Pending,356,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Join_Date,1020,760,6/10/2024,5,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Salary,996.0,NaN,NaN,NaN,85155.056396,19873.727918,50047.32,68392.4875,85547.87,100974.0275,119971.65
Email,1020,64,grace.brown@example.com,27,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Phone,1020.0,NaN,NaN,NaN,-4942252959.221569,2817326049.742002,-9994973202.0,-7341991816.5,-4943996508.5,-2520390678.75,-3896086.0


In [6]:
quality_report = pd.DataFrame({
    "Column": df.columns,
    "Null_Count": df.isnull().sum().values,
    "Unique_Values": df.nunique().values,
    "Data_Type": df.dtypes.astype(str).values
})

quality_report

,Column,Null_Count,Unique_Values,Data_Type
0,Employee_ID,0,1020,str
1,First_Name,0,8,str
2,Last_Name,0,8,str
3,Age,211,4,float64
4,Department_Region,0,36,str
5,Status,0,3,str
6,Join_Date,0,760,str
7,Salary,24,978,float64
8,Email,0,64,str
9,Phone,0,1020,int64


In [7]:
duplicate_count = df.duplicated().sum()

print("Duplicate rows:", duplicate_count)

Duplicate rows: 0


In [8]:
# Check numeric columns for possible range anomalies

print("Age range:")
print("Minimum:", df["Age"].min())
print("Maximum:", df["Age"].max())

print("\nSalary range:")
print("Minimum:", df["Salary"].min())
print("Maximum:", df["Salary"].max())

print("\nPerformance Score values:")
print(df["Performance_Score"].value_counts(dropna=False))

Age range:
Minimum: 25.0
Maximum: 40.0

Salary range:
Minimum: 50047.32
Maximum: 119971.65

Performance Score values:
Performance_Score
Good         270
Average      267
Excellent    267
Poor         216
Name: count, dtype: int64


## Data Quality Report

A data quality assessment was performed on the original dataset before applying the cleaning steps.

The report evaluates the following areas:

- Missing values in each column
- Duplicate records
- Data types
- Basic value-range anomalies in numerical fields

The purpose of this assessment is to identify potential data-quality issues that may affect further analysis and to provide a baseline for the subsequent cleaning process.

In [9]:
# Data Quality Report
# Assess the original dataset before cleaning

quality_df = pd.read_csv("../dataset/Messy_Employee_dataset.csv")

print("=" * 50)
print("DATA QUALITY REPORT")
print("=" * 50)

# 1. Missing values by column
print("\n1. Missing Values by Column")
missing_values = quality_df.isnull().sum()
print(missing_values)

print("\nTotal missing values:", missing_values.sum())

# 2. Duplicate records
print("\n2. Duplicate Records")
duplicate_count = quality_df.duplicated().sum()
print("Total duplicate rows:", duplicate_count)

# 3. Data Types
print("\n3. Data Types")
print(quality_df.dtypes)

# 4. Dataset Dimensions
print("\n4. Dataset Dimensions")
print("Rows:", quality_df.shape[0])
print("Columns:", quality_df.shape[1])

# 5. Numerical Value-Range Check
print("\n5. Numerical Value-Range Check")

print("\nAge range:")
print("Minimum:", quality_df["Age"].min())
print("Maximum:", quality_df["Age"].max())

print("\nSalary range:")
print("Minimum:", quality_df["Salary"].min())
print("Maximum:", quality_df["Salary"].max())

print("\nPerformance Score values:")
print(quality_df["Performance_Score"].value_counts(dropna=False))

DATA QUALITY REPORT

1. Missing Values by Column
Employee_ID            0
First_Name             0
Last_Name              0
Age                  211
Department_Region      0
Status                 0
Join_Date              0
Salary                24
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
dtype: int64

Total missing values: 235

2. Duplicate Records
Total duplicate rows: 0

3. Data Types
Employee_ID              str
First_Name               str
Last_Name                str
Age                  float64
Department_Region        str
Status                   str
Join_Date                str
Salary               float64
Email                    str
Phone                  int64
Performance_Score        str
Remote_Work             bool
dtype: object

4. Dataset Dimensions
Rows: 1020
Columns: 12

5. Numerical Value-Range Check

Age range:
Minimum: 25.0
Maximum: 40.0

Salary range:
Minimum: 50047.32
Maximum: 119971.65

Performance Score

## 2. Missing Data Handling

The dataset contains missing values in the `Age` and `Salary` columns.

### Age
Age has 211 missing values. Since age is a numerical variable and can contain extreme values, median imputation is used. The median is less affected by potential outliers than the mean and provides a robust estimate of the typical employee age.

### Salary
Salary has 24 missing values. Since salary is a numerical variable that may be influenced by high earners, median imputation is also used. This avoids allowing extreme salary values to disproportionately affect the imputed values.

No rows are deleted because the missing values are limited to specific columns and can be reasonably imputed while preserving the available employee records.


In [10]:
# Handle missing values using median imputation

df["Age"] = df["Age"].fillna(df["Age"].median())
df["Salary"] = df["Salary"].fillna(df["Salary"].median())

print("Missing values after imputation:")
print(df[["Age", "Salary"]].isnull().sum())

Missing values after imputation:
Age       0
Salary    0
dtype: int64


### Missing Data Handling Decisions

The missing values were handled using column-specific strategies based on the nature of each variable.

- **Age:** Missing values were replaced using the median because Age is a numerical variable and the median is less affected by extreme values than the mean.
- **Salary:** Missing values were replaced using the median because Salary may contain unusually high or low values, making the median a more robust measure of central tendency.
- **Other columns:** Columns without missing values did not require imputation.

The selected strategies preserve the maximum number of employee records while reducing the effect of extreme values on the imputed results.

## 3. Duplicate Removal

Duplicate rows were checked using the complete dataset.

The dataset contains 0 duplicate rows, so no records were removed.

Keeping this validation step ensures that duplicate employee records do not affect downstream analysis.

In [11]:
# Identify and remove duplicate rows

duplicate_count = df.duplicated().sum()

print("Duplicate rows before removal:", duplicate_count)

df = df.drop_duplicates()

print("Rows after duplicate removal:", len(df))

Duplicate rows before removal: 0
Rows after duplicate removal: 1020


### Duplicate Removal Decision

Duplicate rows were identified by checking the complete dataset.

The dataset contained **0 duplicate rows**.

Therefore, no records were removed during the duplicate-removal step, and the dataset row count remained unchanged.

The duplicate check was still performed to confirm that the dataset did not contain repeated records.

## 4. Data Standardisation

The dataset contains categorical and date fields that require standardisation before analysis.

Categorical values are normalised to consistent labels so that variations in capitalization, spacing, or formatting are treated as the same category.

The `Join_Date` column is converted from text to a proper datetime format to support date-based analysis.

In [12]:
# Standardise categorical columns

# Remove leading/trailing spaces
df["Department_Region"] = df["Department_Region"].str.strip()
df["Status"] = df["Status"].str.strip().str.title()
df["Performance_Score"] = df["Performance_Score"].str.strip().str.title()

# Standardise department/region formatting
df["Department_Region"] = df["Department_Region"].str.replace(
    r"\s*-\s*", "-", regex=True
)

# Convert Join_Date from text to datetime
df["Join_Date"] = pd.to_datetime(df["Join_Date"], errors="coerce")

# Convert Phone to string because phone numbers are identifiers
df["Phone"] = df["Phone"].astype("string")

print("Standardisation completed.")

Standardisation completed.


In [13]:
print("Status values:")
print(df["Status"].value_counts())

print("\nPerformance Score values:")
print(df["Performance_Score"].value_counts())

print("\nJoin_Date dtype:")
print(df["Join_Date"].dtype)

print("\nDepartment/Region sample:")
print(df["Department_Region"].unique()[:10])

Status values:
Status
Pending     356
Active      352
Inactive    312
Name: count, dtype: int64

Performance Score values:
Performance_Score
Good         270
Average      267
Excellent    267
Poor         216
Name: count, dtype: int64

Join_Date dtype:
datetime64[us]

Department/Region sample:
<StringArray>
[  'DevOps-California',       'Finance-Texas',        'Admin-Nevada',
  'Cloud Tech-Florida',         'Sales-Texas',     'DevOps-New York',
 'Cloud Tech-New York',         'HR-New York',  'Finance-California',
       'Sales-Florida']
Length: 10, dtype: str


### Data Standardisation Decisions

The dataset contained inconsistent formatting across categorical and date fields.

The following standardisation steps were applied:

- Categorical values were normalised to consistent representations.
- Leading and trailing whitespace was removed from text fields.
- Department and region values were standardised to consistent formatting.
- Join_Date values were converted from inconsistent text representations into a standard datetime format.

These changes ensure that equivalent values are represented consistently and can be reliably used for analysis and grouping.

## 5. Outlier Detection

Outliers were identified in numerical variables using the Interquartile Range (IQR) method.

The IQR method identifies observations that fall below Q1 - 1.5 × IQR or above Q3 + 1.5 × IQR.

Age and Salary were examined separately because extreme values in these variables may affect downstream analysis and summary statistics.

In [14]:
# Detect outliers using the IQR method

numeric_columns = ["Age", "Salary"]

outlier_summary = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ]

    outlier_summary.append({
        "Column": column,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "Lower_Bound": lower_bound,
        "Upper_Bound": upper_bound,
        "Outlier_Count": len(outliers)
    })

outlier_summary = pd.DataFrame(outlier_summary)

outlier_summary

,Column,Q1,Q3,IQR,Lower_Bound,Upper_Bound,Outlier_Count
0,Age,30.0000,35.0000,5.00,22.5000,42.5000,0
1,Salary,68811.2325,100372.6625,31561.43,21469.0875,147714.8075,0


### Outlier Assessment

The IQR method was used to identify potential outliers in Age and Salary.

No outliers were detected in either variable. All employee records fall within the calculated IQR boundaries.

Therefore, no records were removed or modified based on outlier analysis. Retaining all observations preserves the complete employee population for downstream analysis.

## 6. Final Data Validation

After completing the data-cleaning process, the dataset was validated to ensure that the required cleaning steps were successful.

The final validation checks:
- Remaining missing values
- Duplicate records
- Dataset dimensions
- Data types
- Overall data readiness

The cleaned dataset is considered ready for downstream analysis after these checks are completed.

In [20]:
# Check whether required data types are correct

from pandas.api.types import (
    is_string_dtype,
    is_float_dtype,
    is_datetime64_any_dtype,
    is_bool_dtype
)

# Helper function for text/string columns
def is_text_column(series):
    non_null = series.dropna()
    return non_null.map(lambda x: isinstance(x, str)).all()


dtype_checks = {
    "Employee_ID": is_text_column(df["Employee_ID"]),
    "First_Name": is_text_column(df["First_Name"]),
    "Last_Name": is_text_column(df["Last_Name"]),
    "Age": is_float_dtype(df["Age"]),
    "Department_Region": is_text_column(df["Department_Region"]),
    "Status": is_text_column(df["Status"]),
    "Join_Date": is_datetime64_any_dtype(df["Join_Date"]),
    "Salary": is_float_dtype(df["Salary"]),
    "Email": is_text_column(df["Email"]),
    "Phone": is_text_column(df["Phone"]),
    "Performance_Score": is_text_column(df["Performance_Score"]),
    "Remote_Work": is_bool_dtype(df["Remote_Work"])
}

# Overall dtype validation
dtype_accuracy = all(dtype_checks.values())

print("\nDtype accuracy:", dtype_accuracy)

# Final status
if total_missing == 0 and total_duplicates == 0 and dtype_accuracy:
    print("\nData validation passed.")
    print("Dataset is ready for further analysis.")
else:
    print("\nData validation requires further review.")


Dtype accuracy: True

Data validation passed.
Dataset is ready for further analysis.


### Validation Result

The final validation confirmed that the cleaned dataset contains no remaining missing values and no duplicate records.

The dataset retains 1,020 employee records across 12 columns. The data types were reviewed, and the Join_Date field is stored in a proper datetime format.

The data-cleaning process has therefore been successfully completed, and the dataset is ready for exploratory data analysis (EDA).

## 7. Before vs. After Cleaning Summary

This section compares the original messy dataset with the cleaned dataset to evaluate the impact of the cleaning process.

The comparison includes:

- Row count
- Total null values
- Duplicate rows
- Data type accuracy

The purpose is to demonstrate that the cleaning process improved data quality while preserving valid employee records.

In [21]:
# Load the original messy dataset for before-vs-after comparison

df_before = pd.read_csv("../dataset/Messy_Employee_dataset.csv")

# Before cleaning metrics
before_rows = len(df_before)
before_missing = df_before.isnull().sum().sum()
before_duplicates = df_before.duplicated().sum()


# Check expected data types using semantic type checks
def check_data_types(dataframe):

    checks = {
        "Employee_ID": is_text_column(dataframe["Employee_ID"]),
        "First_Name": is_text_column(dataframe["First_Name"]),
        "Last_Name": is_text_column(dataframe["Last_Name"]),
        "Age": is_float_dtype(dataframe["Age"]),
        "Department_Region": is_text_column(dataframe["Department_Region"]),
        "Status": is_text_column(dataframe["Status"]),
        "Join_Date": is_datetime64_any_dtype(dataframe["Join_Date"]),
        "Salary": is_float_dtype(dataframe["Salary"]),
        "Email": is_text_column(dataframe["Email"]),
        "Phone": is_text_column(dataframe["Phone"]),
        "Performance_Score": is_text_column(dataframe["Performance_Score"]),
        "Remote_Work": is_bool_dtype(dataframe["Remote_Work"])
    }

    return sum(checks.values())


# Count dtype matches
before_dtype_accuracy = check_data_types(df_before)
after_dtype_accuracy = check_data_types(df)


# After cleaning metrics
after_rows = len(df)
after_missing = df.isnull().sum().sum()
after_duplicates = df.duplicated().sum()


# Create comparison table
comparison = pd.DataFrame({
    "Metric": [
        "Row Count",
        "Total Missing Values",
        "Duplicate Rows",
        "Correct Data Types"
    ],
    "Before Cleaning": [
        before_rows,
        before_missing,
        before_duplicates,
        f"{before_dtype_accuracy}/{len(df.columns)}"
    ],
    "After Cleaning": [
        after_rows,
        after_missing,
        after_duplicates,
        f"{after_dtype_accuracy}/{len(df.columns)}"
    ]
})

comparison

,Metric,Before Cleaning,After Cleaning
0,Row Count,1020,1020
1,Total Missing Values,235,0
2,Duplicate Rows,0,0
3,Correct Data Types,10/12,12/12


### Summary

The before-vs-after comparison confirms the impact of the data-cleaning process.

Missing values and duplicate records were successfully addressed, while the original employee records were preserved. Data types were also corrected to ensure that identifiers, dates, numeric values, and categorical fields are stored appropriately.

The resulting dataset is clean, consistent, and ready for exploratory data analysis (EDA).

# 8. Save Cleaned Dataset

The cleaned dataset is saved as a separate CSV file so that the validated data can be used for downstream exploratory analysis without modifying the original dataset.

The cleaned dataset preserves the valid employee records after the data-cleaning process.

In [17]:
# Save the cleaned dataset for further analysis

cleaned_file_path = "../outputs/cleaned_employee_dataset.csv"

df.to_csv(cleaned_file_path, index=False)

print("Cleaned dataset saved successfully.")
print("File:", cleaned_file_path)

Cleaned dataset saved successfully.
File: ../outputs/cleaned_employee_dataset.csv
